app/models/spree/order/checkout.rb

```diff
@@ -75,9 +75,9 @@ def self.define_state_machine!
 
               before_transition to: :delivery, do: :create_proposed_shipments
               before_transition to: :delivery, do: :ensure_available_shipping_rates
-              before_transition to: :payment, do: :create_tax_charge!
               before_transition to: :confirmation, do: :validate_payment_method!
 
+              after_transition to: :payment, do: :create_tax_charge!
```

Comment:
if we keep the original code  `before_transition`, `create_tax_charge!` will do nothing because of the guard clause.
```ruby
    def create_tax_charge!
      return if state.in?(["cart", "address", "delivery"])
...
    end
```
It was introduced on this [PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/8810) for optimization reasons.

Reply:
Taxing the items in the order needs to happen after the order's addresses have been set but should be done before the user lands on the payments step page. It wasn't quite working correctly before as the order's state hadn't been updated at the point it was being checked in the `before_transition` hook. The `after_transition` hook is the correct place for it :+1:

Reply:
Thanks. It would be cool if we could add a spec to show that, but the main thing is we're improving 👍